### PR TITLE
EZP-30997: Fixed permission checks when copying translations during publishing

### PR DESCRIPTION
**Pull request by pawbuj**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-30997](https://jira.ez.no/browse/EZP-30997)
| **Bug/Improvement**| yes
| **New feature**    | no
| **Target version** |`7.5` 
| **BC breaks**      | no
| **Tests pass**     | no
| **Doc needed**     | no

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 1]
```diff
@@ -6424,4 +6425,50 @@ private function createContentWithReverseRelations(array $drafts)
 
         return $contentWithReverseRelations;
     }
+
+    public function testPublishVersionByUserWithLanguageLimitation()
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: Please:
1. Move this test to a dedicated test case, e.g.: `\eZ\Publish\API\Repository\Tests\Limitation\LanguageLimitationTest::testPublishVersionTranslation`.

2. Add `: void` return type.

3. Create a negative test case - publishing explicit translations list, but for a user who shouldn't be allowed to do so (e.g. trying to publish English, while having limitation to German).

File: eZ/Publish/API/Repository/Tests/ContentServiceTest.php [hunk 2]
```diff
@@ -6424,4 +6425,50 @@ private function createContentWithReverseRelations(array $drafts)
...
+                self::GER_DE => 'Published DE',
+            ],
+            $this->generateId('location', 2)
+        );
+
+        $draft = $this->contentService->createContentDraft($publishedContent->contentInfo);
+
+        $contentUpdateStruct = new ContentUpdateStruct([
+            'initialLanguageCode' => self::GER_DE,
+        ]);
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/ContentServiceTest.php]: `ContentUpdateStruct` is not a part of SPI, so end-developer should not instantiate it directly.
What we should do here, to advocate good practice, is:
```php
$contentUpdateStruct = $contentService->newContentUpdateStruct();
$contentUpdateStruct->initialLanguageCode = self::GER_DE;
```

File: eZ/Publish/Core/Repository/ContentService.php [hunk 3]
```diff
@@ -1929,6 +1955,18 @@ public function loadRelations(APIVersionInfo $versionInfo)
             throw new UnauthorizedException('content', $function);
         }
 
+        return $this->internalLoadRelations($versionInfo);
+    }
+
+    /**
+     * Loads all outgoing relations for the given version without checking the permissions.
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException if the user is not allowed to read this version
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentService.php]: This is no longer true, what is true though:

```suggestion
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
```

File: eZ/Publish/Core/Repository/ContentService.php [hunk 4]
```diff
@@ -1670,7 +1696,7 @@ protected function copyTranslationsFromPublishedVersion(APIVersionInfo $versionI
             }
         }
 
-        $this->updateContent($versionInfo, $updateStruct);
+        $this->internalUpdateContent($versionInfo, $updateStruct);
```
Comment (alongosz) [inline eZ/Publish/Core/Repository/ContentService.php]: This change requires `copyTranslationsFromPublishedVersion` PHPDoc to be updated.

File: eZ/Publish/API/Repository/Tests/Limitation/LanguageLimitationTest.php [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\API\Repository\Tests\Limitation;
+
+use eZ\Publish\API\Repository\Exceptions\UnauthorizedException;
+use eZ\Publish\API\Repository\Tests\BaseTest;
+use eZ\Publish\API\Repository\Values\User\Limitation\LanguageLimitation;
+
+class LanguageLimitationTest extends BaseTest
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Limitation/LanguageLimitationTest.php]: I'm sorry, I've meant you should add test case to the file `eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php`. Seems like namespace is wrong there and I didn't notice it when pasting it. The namespace needs to be fixed as well.

**Comment by alongosz** [comment]
While reviewing once again #2846 I've discovered yet another bug here. Consider the following use case:
```php
public function testPublishVersionTranslationIsNotAllowed(): void
{
    $repository = $this->getRepository();
    $contentService = $repository->getContentService();
    $permissionResolver = $repository->getPermissionResolver();

    $britishEditor = $this->createEditorUserWithLanguageLimitation(['eng-GB'], 'editor-uk');
    $americanEditor = $this->createEditorUserWithLanguageLimitation(['eng-US'], 'editor-us');

    // British editor publishes BrE content
    $permissionResolver->setCurrentUserReference($britishEditor);

    $folder = $this->createFolder(['eng-GB' => 'BrE Folder'], 2);

    // American editor creates and saves AmE draft
    $permissionResolver->setCurrentUserReference($americanEditor);

    $folder = $contentService->loadContent($folder->id);
    $folderDraft = $contentService->createContentDraft($folder->contentInfo);
    $folderUpdateStruct = $contentService->newContentUpdateStruct();
    $folderUpdateStruct->setField('name', 'AmE Folder', 'eng-US');
    $folderDraft = $contentService->updateContent(
        $folderDraft->versionInfo,
        $folderUpdateStruct
    );

    // British editor tries to publish AmE translation
    $permissionResolver->setCurrentUserReference($britishEditor);
    $folderDraftVersionInfo = $contentService->loadVersionInfo(
        $folderDraft->contentInfo,
        $folderDraft->versionInfo->versionNo
    );
    self::assertTrue($folderDraftVersionInfo->isDraft());
    $this->expectException(UnauthorizedException::class);
    $this->expectExceptionMessage("User does not have access to 'publish' 'content'");
    $contentService->publishVersion($folderDraftVersionInfo, ['eng-US']);
}

private function createEditorUserWithLanguageLimitation(
    array $allowedTranslationsList,
    string $login = 'editor'
): User {
    $limitations = [
        // limitation for specific translations
        new LanguageLimitation(['limitationValues' => $allowedTranslationsList]),
    ];

    return $this->createUserWithPolicies(
        $login,
        [
            ['module' => 'content', 'function' => 'read'],
            ['module' => 'content', 'function' => 'versionread'],
            ['module' => 'content', 'function' => 'view_embed'],
            ['module' => 'content', 'function' => 'create', 'limitations' => $limitations],
            ['module' => 'content', 'function' => 'edit', 'limitations' => $limitations],
            ['module' => 'content', 'function' => 'publish', 'limitations' => $limitations],
        ]
    );
}
``` 
I'm not sure if this is reproducible with AdminUI (depends if ATM it's possible to edit someone else's draft - don't remember). However for the API it's an issue.

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 6]
```diff
@@ -235,4 +249,136 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
         $this->expectException(UnauthorizedException::class);
         $contentService->publishVersion($folderDraft->getVersionInfo());
     }
+
+    public function testPublishVersionTranslation(): void
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]: Nitpick: I prefer having exceptions documented, so IDE doesn't report them as not handled:
```suggestion
    /**
     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
     */
    public function testPublishVersionTranslation(): void
```

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 7]
```diff
@@ -235,4 +249,136 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
...
+        $this->assertEquals(
+            [
+                self::ENG_US => 'Published US',
+                self::GER_DE => 'Draft 1 DE',
+            ],
+            $content->fields['name']
+        );
+    }
+
+    public function testPublishVersionTranslationIsNotAllowed(): void
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]:
```suggestion
    /**
     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
     */
    public function testPublishVersionTranslationIsNotAllowed(): void
```

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 8]
```diff
@@ -235,4 +249,136 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
...
+
+        $contentService->updateContent($draft->versionInfo, $contentUpdateStruct);
+
+        $permissionResolver->setCurrentUserReference($this->createEditorUserWithLanguageLimitation(['ger-DE']));
+
+        $this->expectException(UnauthorizedException::class);
+        $contentService->publishVersion($draft->versionInfo, [self::ENG_US]);
+    }
+
+    public function testPublishVersionTranslationIsNotAllowedWithTwoEditors(): void
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]:
```suggestion
    /**
     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
     */
    public function testPublishVersionTranslationIsNotAllowedWithTwoEditors(): void
```

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 9]
```diff
@@ -235,4 +249,136 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
...
+            $folderDraft->contentInfo,
+            $folderDraft->versionInfo->versionNo
+        );
+        self::assertTrue($folderDraftVersionInfo->isDraft());
+        $this->expectException(UnauthorizedException::class);
+        $this->expectExceptionMessage("User does not have access to 'publish' 'content'");
+        $contentService->publishVersion($folderDraftVersionInfo, [self::ENG_US]);
+    }
+
+    public function testPublishVersionTranslationWhenUserHasNoAccessToAllLanguages(): void
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]:
```suggestion
    /**
     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
     */
    public function testPublishVersionTranslationWhenUserHasNoAccessToAllLanguages(): void
```

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php [hunk 10]
```diff
@@ -126,4 +126,26 @@ function (Field $field) {
...
+    /**
+     * Set intent to publish, to specified languages, all from the given list.
+     *
+     * @param array $languageCodes
+     *
+     * @return self
+     *
+     * @throws \eZ\Publish\API\Repository\Exceptions\InvalidArgumentException
+     */
+    public function publishLanguages(array $languageCodes): self
```
Comment (alongosz) [inline eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php]: We're not publishing a language but a translation.
```suggestion
    public function publishTranslations(array $languageCodes): self
```

**Comment by alongosz** [comment]
@mikadamczyk one more request - could you incorporate #2846 here as well, as it touches the same portion of code? Would be easier to test.

**Comment by mikadamczyk** [comment]
Sure @alongosz I thought that they ware divided by some purpose, but definitely it touches the same part of code

**Comment by alongosz** [comment]
> Sure @alongosz I thought that they were divided by some purpose, but definitely it touches the same part of code

Yes, initially they were, but given the changes here it no longer makes sense :)

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 11]
```diff
@@ -235,4 +249,156 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
...
+        $this->expectException(UnauthorizedException::class);
+        $contentService->publishVersion($draft->versionInfo, [self::ENG_US]);
+    }
+
+    /**
+     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
+     */
+    public function testPublishVersionTranslationIsNotAllowedWithTwoEditors(): void
```
Comment (ViniTou) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]: you are missing const usage for languages codes here

File: eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php [hunk 12]
```diff
@@ -126,4 +126,26 @@ function (Field $field) {
 
         return $this;
     }
+
+    /**
+     * Set intent to publish, to specified translations, all from the given list.
+     *
+     * @param array $languageCodes
```
Comment (adamwojs) [inline eZ/Publish/SPI/Limitation/Target/Builder/VersionBuilder.php]:
```suggestion
     * @param string[] $languageCodes
```

File: eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php [hunk 13]
```diff
@@ -235,4 +249,156 @@ public function testPublishVersionIsNotAllowedIfModifiedOtherTranslations(array
...
+        $contentService->publishVersion($draft->versionInfo, [self::GER_DE, self::ENG_GB]);
+    }
+
+    /**
+     * @param \eZ\Publish\API\Repository\ContentService $contentService
+     *
+     * @return \eZ\Publish\API\Repository\Values\Content\Content
+     * @throws \eZ\Publish\API\Repository\Exceptions\ForbiddenException
+     * @throws \eZ\Publish\API\Repository\Exceptions\NotFoundException
+     * @throws \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Tests/Values/User/Limitation/LanguageLimitationTest.php]: Nitpick: The newline is missing between `@return` and `throws`

**Pull request merged**